### PR TITLE
[CPU] FullyConnected, Gemm node migration on nGraph

### PR DESCRIPTION
**Pull request by mandrono**
Don't review part associated with ScaleShift transformation, will be fixed after merge https://github.com/dmitry-gorokhov/openvino/pull/23

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp [hunk 1]
```diff
@@ -50,84 +77,37 @@ std::vector<memory::format_tag> MKLDNNFullyConnectedNode::getAvailableFormatsFor
...
-//        MKLDNNMemoryDesc out_candidate(outDims, outputDataType, memory::format_tag::any);
-//
-//        createDescriptor({in_candidate}, {out_candidate});
-//    }
+    if (getParentEdges().size() != 3)
+        THROW_IE_EXCEPTION << errorPrefix << " has incorrect number of input edges";
+    if (getChildEdges().empty())
+        THROW_IE_EXCEPTION<< errorPrefix << " has incorrect number of output edges";
+
+    auto inputDataType = MKLDNNExtensionUtils::IEPrecisionToDataType(getOriginalInputPrecisions()[DATA_ID]);
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp:85]: You know what to do

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_gemm_node.cpp [hunk 2]
```diff
@@ -12,44 +12,70 @@
 #include <mkldnn_extension_utils.h>
 #include "ie_parallel.hpp"
 #include "common/cpu_memcpy.h"
+#include <ngraph/opsets/opset1.hpp>
 
 using namespace mkldnn;
 using namespace MKLDNNPlugin;
 using namespace InferenceEngine;
 
+bool MKLDNNGemmNode::isSupportedOperation(const std::shared_ptr<ngraph::Node>& op, std::string& errorMessage) noexcept {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_gemm_node.cpp:21]: Lets rename  `Gemm` -> `Matmul`

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp [hunk 3]
```diff
@@ -17,20 +17,47 @@ using namespace mkldnn;
...
-//    if (getCnnLayer()->type == "FullyConnected" || getCnnLayer()->type == "InnerProduct") {
-//        baseInputsNumber = getCnnLayer().get()->insData.size();
-//    }
+        : MKLDNNNode(op, eng, cache), withBiases(false) {
+    std::string errorMessage;
+    if (isSupportedOperation(op, errorMessage)) {
+        errorPrefix = "FullyConnected node with name '" + getName() + "'";
+
+        const auto bias = std::dynamic_pointer_cast<const ngraph::opset1::Constant>(op->get_input_node_shared_ptr(BIAS_ID))->cast_vector<float>();
+        withBiases = std::any_of(bias.begin(), bias.end(), [](float b){ return b != 0.0f; });
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp:56]: Cannot we just don't create bias input during transformation if all values are zero?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp [hunk 4]
```diff
@@ -17,20 +17,47 @@ using namespace mkldnn;
...
+        }
+        if (std::dynamic_pointer_cast<const ngraph::opset1::Constant>(fc->get_input_node_shared_ptr(WEIGHTS_ID)) == nullptr) {
+            errorMessage = "Only Constant operation on 'weights' input is supported";
+            return false;
+        }
+        if (std::dynamic_pointer_cast<const ngraph::opset1::Constant>(fc->get_input_node_shared_ptr(BIAS_ID)) == nullptr) {
+            errorMessage = "Only Constant operation on 'bias' input is supported";
+            return false;
+        }
+        if (fc->get_input_shape(DATA_ID).size() != 2) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp:35]: We still need to keep 3d and 4d support

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp [hunk 5]
```diff
@@ -687,110 +686,41 @@ void MKLDNNGraphOptimizer::MergeTwoEqualScaleShifts(MKLDNNGraph& graph) {
...
-//                removeEdge(graph, p_edge);
-//            }
-//        }
-//
-//        graph.DropNode(childNode);
-//    }
+    auto& graphNodes = graph.GetNodes();
+
+    auto isSutableParentNode = [](MKLDNNNodePtr node) {
+        return node->getType() == FullyConnected && node->getChildEdges().size() == 1;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp:692]: We need to disable fusing into 3D FC

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 6]
```diff
@@ -282,6 +283,8 @@ static void Transformation(CNNNetwork& clonedNetwork, const Config& conf) {
         return node->get_rt_info().count("UNROLL_TI") == 0;
     });
 
+    ConvertToLegacyOpset(nGraphFunc);
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp (hunk 1 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp]: done

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_gemm_node.cpp (hunk 2 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_gemm_node.cpp]: renamed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp (hunk 3 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp]: transformations was changed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp (hunk 4 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_fullyconnected_node.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp (hunk 5 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp]: fixed

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**